Error while checking build/O2/fullCI for 8d711470ff0ea941a21a4945e59ed71d80283f53 at 2022-10-03 02:27:
```
## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9932-slc8_x86-64/0/Detectors/TOF/calibration/src/TOFChannelCalibrator.cxx:23:10: error: 'omp.h' file not found [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/9932/8d711470ff0ea941a21a4945e59ed71d80283f53/build_O2_fullCI/pretty.html).

--- a/Detectors/CPV/calib/include/CPVCalibration/GainCalibrator.h
+++ b/Detectors/CPV/calib/include/CPVCalibration/GainCalibrator.h
@@ -42,7 +42,8 @@ class AmplitudeSpectrum
   double getRMS() const { return (mSumA2 / mNEntries) - ((mSumA * mSumA) / (mNEntries * mNEntries)); }; // return RMS of distribution
   uint32_t getNEntries() const { return mNEntries; }
   const uint32_t* getBinContent() { return mBinContent.data(); } // since C++17 std::array::data is constexpr
-  void fillBinData(TH1F* h);
+  void dumpToHisto(TH1F* h);
+  int nEventsInRange(float lR, float rR);
 
  private:
   uint32_t mNEntries;

--- a/Detectors/CPV/calib/src/GainCalibrator.cxx
+++ b/Detectors/CPV/calib/src/GainCalibrator.cxx
@@ -58,7 +58,7 @@ AmplitudeSpectrum& AmplitudeSpectrum::operator+=(const AmplitudeSpectrum& rhs)
 void AmplitudeSpectrum::fill(float amplitude)
 {
   if ((lRange <= amplitude) && (amplitude < rRange)) {
-    int bin = (amplitude - lRange) / nBins;
+    int bin = ((amplitude - lRange) / (rRange - lRange)) * nBins;
     mBinContent[bin]++;
     mNEntries++;
     mSumA += amplitude;
@@ -66,12 +66,34 @@ void AmplitudeSpectrum::fill(float amplitude)
   }
 }
 //_____________________________________________________________________________
-void AmplitudeSpectrum::fillBinData(TH1F* h)
+void AmplitudeSpectrum::dumpToHisto(TH1F* h)
 {
-  for (uint16_t i = 0; i < nBins; i++) {
-    h->SetBinContent(i + 1, float(mBinContent[i]));
-    h->SetBinError(i + 1, sqrt(float(mBinContent[i])));
+  int rebin = nBins / h->GetNbinsX();
+  if (h != nullptr) {
+    for (int iBin = 0; iBin < h->GetNbinsX(); iBin++) {
+      float binC = 0.;
+      for (uint16_t i = iBin * rebin; i < (iBin + 1) * rebin; i++) {
+        binC += mBinContent[i];
+      }
+      h->SetBinContent(iBin + 1, binC);
+    }
   }
+}
+int AmplitudeSpectrum::nEventsInRange(float lR, float rR)
+{
+  if (lR < lRange) {
+    lR = lRange;
+  }
+  if (rR > rRange) {
+    rR = rRange;
+  }
+  int first = ((lR - lRange) / (rRange - lRange)) * nBins;
+  int last = ((rR - lRange) / (rRange - lRange)) * nBins;
+  int nEvents = 0;
+  for (int i = first; i < last; i++) {
+    nEvents += mBinContent[i];
+  }
+  return nEvents;
 }
 //_____________________________________________________________________________
 // GainCalibData
@@ -127,6 +149,8 @@ void GainCalibrator::configParameters()
   mMaxAllowedCoeff = cpvParams.gainMaxAllowedCoeff;
   mFitRangeL = cpvParams.gainFitRangeL;
   mFitRangeR = cpvParams.gainFitRangeR;
+  mUpdateTFInterval = cpvParams.gainCheckForCalibrationInterval;
+
   // adjust fit ranges to descrete binned values
   if (mFitRangeL < AmplitudeSpectrum::lRange) {
     mFitRangeL = AmplitudeSpectrum::lRange;
@@ -134,7 +158,7 @@ void GainCalibrator::configParameters()
   if (mFitRangeR > AmplitudeSpectrum::rRange) {
     mFitRangeR = AmplitudeSpectrum::rRange;
   }
-  double binWidth = (AmplitudeSpectrum::rRange - AmplitudeSpectrum::lRange) / AmplitudeSpectrum::nBins;
+  float binWidth = (AmplitudeSpectrum::rRange - AmplitudeSpectrum::lRange) / AmplitudeSpectrum::nBins;
   mFitRangeL = AmplitudeSpectrum::lRange + std::floor((mFitRangeL - AmplitudeSpectrum::lRange) / binWidth) * binWidth;
   mFitRangeR = AmplitudeSpectrum::lRange + std::floor((mFitRangeR - AmplitudeSpectrum::lRange) / binWidth) * binWidth;
 
@@ -146,6 +170,7 @@ void GainCalibrator::configParameters()
   LOG(info) << "mMaxAllowedCoeff = " << mMaxAllowedCoeff;
   LOG(info) << "mFitRangeL = " << mFitRangeL;
   LOG(info) << "mFitRangeR = " << mFitRangeR;
+  LOG(info) << "mUpdateTFInterval =  " << mUpdateTFInterval;
 }
 //_____________________________________________________________________________
 void GainCalibrator::initOutput()
@@ -170,21 +195,27 @@ void GainCalibrator::finalizeSlot(GainTimeSlot& slot)
   TF1* fLandau = new TF1("fLandau", "landau", mFitRangeL, mFitRangeR);
   fLandau->SetParLimits(0, 0., 1.E6);
   fLandau->SetParLimits(1, mDesiredLandauMPV / mMaxAllowedCoeff, mDesiredLandauMPV / mMinAllowedCoeff);
-  fLandau->SetParLimits(1, 0., 1.E3);
+  fLandau->SetParLimits(2, 0., 1.E3);
   TH1F h("histoCPVMaxAmplSpectrum", "", AmplitudeSpectrum::nBins, AmplitudeSpectrum::lRange, AmplitudeSpectrum::rRange);
+  h.Rebin(std::ceil(mMaxAllowedCoeff)); // rebin histogram in order to avoid descrete structures in it
   // double binWidth = (AmplitudeSpectrum::rRange - AmplitudeSpectrum::lRange) / AmplitudeSpectrum::nBins;
   // size_t nBinsToFit = (mFitRangeR - mFitRangeL) / binWidth;
   // uint32_t xMin = (mFitRangeL - AmplitudeSpectrum::lRange) / binWidth;
   // uint32_t xMax = (mFitRangeR - AmplitudeSpectrum::lRange) / binWidth;
   int nCalibratedChannels = 0;
-  for (int i = 0; i < Geometry::kNCHANNELS; i++) {
+  int badChi2Channels = 0;
+  for (unsigned short i = 0; i < Geometry::kNCHANNELS; i++) {
+    newGains->setGain(i, mPreviousGains.get()->getGain(i)); // copy previous gains first
     // print some info
     if ((i % 500) == 0) {
       LOG(info) << "GainCalibrator::finalizeSlot() : checking channel " << i;
     }
-    if (slot.getContainer()->mAmplitudeSpectra[i].getNEntries() > mMinEvents) { // we are ready to fit
+    if (slot.getContainer()->mAmplitudeSpectra[i].getNEntries() > mMinEvents) {                            // we are ready to fit
+      if (slot.getContainer()->mAmplitudeSpectra[i].nEventsInRange(mFitRangeL, mFitRangeR) < mMinEvents) { // actually not enough events in fit range
+        continue;
+      }
       h.Reset();
-      slot.getContainer()->mAmplitudeSpectra[i].fillBinData(&h);
+      slot.getContainer()->mAmplitudeSpectra[i].dumpToHisto(&h);
       // set some starting values
       double mean = slot.getContainer()->mAmplitudeSpectra[i].getMean();
       double rms = slot.getContainer()->mAmplitudeSpectra[i].getRMS();
@@ -193,10 +224,17 @@ void GainCalibrator::finalizeSlot(GainTimeSlot& slot)
       auto fitResult = h.Fit(fLandau, "SQL0N", "", mFitRangeL, mFitRangeR);
       // auto fitResult = o2::math_utils::fit<uint32_t>(nBinsToFit, &(slot.getContainer()->mAmplitudeSpectra[i].getBinContent()[xMin]), xMin, xMax, *fLandau);
       if (fitResult->Chi2() / fitResult->Ndf() > mToleratedChi2PerNDF) {
-        //  in case of bad fit -> do something sofisticated. but what?
+        badChi2Channels++;
+        // in case of bad fit -> do something sofisticated. but what?
         // continue;
-        LOG(info) << "GainCalibrator::finalizeSlot() : bad chi2/ndf in fit of spectrum in channel " << i;
-        fitResult->Print("V");
+        if (badChi2Channels < 20) {
+          LOG(info) << "GainCalibrator::finalizeSlot() : bad chi2/ndf in fit of spectrum in channel " << i;
+          fitResult->Print("V");
+        } else if (badChi2Channels == 20) {
+          LOG(info) << "GainCalibrator::finalizeSlot() : bad chi2/ndf in fit of spectrum in channel " << i;
+          fitResult->Print("V");
+          LOG(info) << "GainCalibrator::finalizeSlot() : bad chi2/ndf is reported 20 times. Muting ";
+        }
       }
       // calib coeffs are defined as mDesiredLandauMPV/actualMPV*previousCoeff
       float coeff = mDesiredLandauMPV / fLandau->GetParameter(1) * mPreviousGains.get()->getGain(i);
@@ -213,7 +251,7 @@ void GainCalibrator::finalizeSlot(GainTimeSlot& slot)
   }
   delete fLandau;
   LOG(info) << "GainCalibrator::finalizeSlot() : succesfully calibrated " << nChannelsCalibrated << "channels";
-
+  LOG(info) << "GainCalibrator::finalizeSlot() : bad chi2/ndf is occured " << badChi2Channels << "times";
   // prepare new ccdb entries for sending
   mGainsVec.push_back(*newGains);
   // metadata for o2::cpv::CalibParams

--- a/Detectors/CPV/calib/testWorkflow/GainCalibratorSpec.h
+++ b/Detectors/CPV/calib/testWorkflow/GainCalibratorSpec.h
@@ -38,21 +38,18 @@ class CPVGainCalibratorSpec : public o2::framework::Task
   void init(o2::framework::InitContext& ic) final
   {
     o2::base::GRPGeomHelper::instance().setRequest(mCCDBRequest);
-    // auto slotL = ic.options().get<uint32_t>("tf-per-slot");
-    // auto delay = ic.options().get<uint32_t>("max-delay");
     auto updateInterval = ic.options().get<uint32_t>("updateInterval"); // in TF
     bool updateAtTheEndOfRunOnly = ic.options().get<bool>("updateAtTheEndOfRunOnly");
     mCalibrator = std::make_unique<o2::cpv::GainCalibrator>();
-    mCalibrator->setSlotLength(0);
-    mCalibrator->setMaxSlotsDelay(1000);
+    mCalibrator->setSlotLength(0); // infinite TF slot
+    mCalibrator->setMaxSlotsDelay(10000);
     if (updateAtTheEndOfRunOnly) {
       mCalibrator->setUpdateAtTheEndOfRunOnly();
     }
     mCalibrator->setCheckIntervalInfiniteSlot(updateInterval);
-    mCalibrator->setUpdateTFInterval(updateInterval);
     LOG(info) << "CPVGainCalibratorSpec initialized";
-    LOG(info) << "tf-per-slot = 0 (this calibrator works only in single infinite slot mode)";
-    LOG(info) << "max-delay = 1000 (inconfigurable for this calibrator)";
+    LOG(info) << "tf-per-slot = 0 (inconfigurable, this calibrator works only in single infinite slot mode)";
+    LOG(info) << "max-delay = 10000 (inconfigurable for this calibrator)";
     LOG(info) << "updateInterval = " << updateInterval;
     LOG(info) << "updateAtTheEndOfRunOnly = " << updateAtTheEndOfRunOnly;
   }

